{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "30bbcddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from selenium import webdriver\n",
    "import time\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from datetime import datetime\n",
    "from selenium.webdriver.support.ui import Select\n",
    "import re\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "import os\n",
    "import zipfile\n",
    "import shutil\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0ef86b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\EHOUT\\AppData\\Local\\Temp/ipykernel_19156/2679306720.py:4: DeprecationWarning: use options instead of chrome_options\n",
      "  driver = webdriver.Chrome(\"C://project_IP//chromedriver.exe\",chrome_options=chrome_options)\n"
     ]
    }
   ],
   "source": [
    "chrome_options = webdriver.ChromeOptions()\n",
    "chrome_options.add_argument(\"--incognito\")\n",
    "\n",
    "driver = webdriver.Chrome(\"C://project_IP//chromedriver.exe\",chrome_options=chrome_options)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2468b2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def exp_to_short(amino_acid):\n",
    "    ''' \n",
    "      A function which can convert expanded forms\n",
    "      like Ala to A\n",
    "    '''\n",
    "\n",
    "    dicts={'Ala':'a', 'Asx':'b', 'Cys':'c',\n",
    "           'Asp':'d', 'Glu':'e', 'Phe':'f',\n",
    "           'Gly':'g', 'His':'h', 'Ile':'i',\n",
    "           'Lys':'k', 'Leu':'l', 'Met':'m',\n",
    "           'Asn':'n', 'Pro':'p', 'Gln':'q',\n",
    "           'Arg':'r', 'Ser':'s', 'Thr':'t',\n",
    "           'Sec':'u', 'Val':'v', 'Trp':'w',\n",
    "           'Xaa':'x', 'Tyr':'y', 'Glx':'z'}\n",
    "    \n",
    "    return dicts[amino_acid]\n",
    "\n",
    "\n",
    "\n",
    "def translate(sequence):\n",
    "    coding_dna = Seq(sequence)\n",
    "    prt=str(coding_dna.translate()) #translate dna to protein\n",
    "    return prt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "148755dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sequences from sequence listing text file\n",
    "def get_data(file, upper=True):\n",
    "    file_content=file\n",
    "    \n",
    "        \n",
    "    seq_no=[]\n",
    "    seq_len=[]\n",
    "    seq_type=[]\n",
    "    organism=[]\n",
    "    other_info=[]\n",
    "    trnsl_seq=[]\n",
    "    \n",
    "    chk=\"<400>\" #to check the previous label\n",
    "    for i in range(len(file_content)):\n",
    "        x=file_content[i].split()\n",
    "        \n",
    "        #get sequence number\n",
    "        if(len(x)>0 and x[0]=='<210>' and chk=='<400>'):\n",
    "            seq_no.append(file_content[i][6:len(file_content[i])])\n",
    "            chk=x[0]\n",
    "            \n",
    "        #get sequence length\n",
    "        if(len(x)>0 and x[0]=='<211>' and chk=='<210>'):\n",
    "            seq_len.append(file_content[i][6:len(file_content[i])])\n",
    "            chk=x[0]\n",
    "        \n",
    "        #get sequence type\n",
    "        if(len(x)>0 and x[0]=='<212>' and chk=='<211>'):\n",
    "            seq_type.append(file_content[i][6:len(file_content[i])])\n",
    "            chk=x[0]\n",
    "            \n",
    "            \n",
    "        #get organism name\n",
    "        if(len(x)>0 and x[0]=='<213>' and chk=='<212>'):\n",
    "            organism.append(file_content[i][6:len(file_content[i])])\n",
    "            chk=x[0]\n",
    "            \n",
    "            \n",
    "        #get other information\n",
    "        if(len(x)>0 and x[0]=='<223>' and chk=='<213>'):\n",
    "            other_info.append(file_content[i][6:len(file_content[i])])\n",
    "            chk=x[0]\n",
    "            \n",
    "        # if any attribute is missing add none to the list\n",
    "        if(len(x)>0 and x[0]=='<400>'):\n",
    "            if(len(other_info)!=len(seq_no)):\n",
    "                #print(len(seq_no), len(other_info))\n",
    "                \n",
    "                other_info.append(\" \")\n",
    "            \n",
    "            if(len(seq_len)!=len(seq_no)):\n",
    "                seq_len.append(0)\n",
    "                \n",
    "            if(len(seq_type)!=len(seq_no)):\n",
    "                seq_type.append(\"None\")\n",
    "                \n",
    "            if(len(organism)!=len(seq_no)):\n",
    "                organism.append(\"None\")\n",
    "                \n",
    "            chk=x[0]\n",
    "            \n",
    "    \n",
    "    seq_code=[] #getting the sequence code    \n",
    "    i=0    \n",
    "    while(i<len(file_content)):\n",
    "        x=file_content[i].split()\n",
    "        if(len(x)>0 and x[0]=='<400>'):\n",
    "            i+=1\n",
    "            \n",
    "            temp=\"\"\n",
    "            \n",
    "            while(x[0]!='<210>' and i<len(file_content)): #append all sequence between 400 and 210\n",
    "                x=file_content[i].split()\n",
    "                if(len(x)>0):\n",
    "                    for j in range(len(x)):\n",
    "                        #convert expanded forms like 'Ala' to 'a'\n",
    "                        try:    \n",
    "                            if(upper==True):\n",
    "                                temp+=exp_to_short(x[j]).upper()\n",
    "                            else:\n",
    "                                temp+=exp_to_short(x[j])\n",
    "                        except:\n",
    "                            # don't add the digits in sequence \n",
    "                            if(x[j].isdigit()==False and x[0]!='<210>'):\n",
    "                               temp+=x[j]\n",
    "                else:\n",
    "                    x.append('0') #to nullify the exception\n",
    "                i+=1\n",
    "                \n",
    "            sz=len(seq_code)\n",
    "            if(seq_type[sz][-3:]!=\"PRT\"): #if sequence type is not protein\n",
    "                prt=translate(temp) \n",
    "                trnsl_seq.append(prt)\n",
    "            else:\n",
    "                trnsl_seq.append(temp)\n",
    "                \n",
    "            seq_code.append(temp) \n",
    "            \n",
    "        else:\n",
    "            i+=1\n",
    "               \n",
    "             \n",
    "    df_dict={'Sequence_ID_number': seq_no, 'Sequence_length':seq_len,\n",
    "             'Sequence_type':seq_type, 'Organism':organism,\n",
    "             'Other_info': other_info, 'Sequence':seq_code,\n",
    "             'Protein': trnsl_seq}\n",
    "    \n",
    "    \n",
    "    df=pd.DataFrame(df_dict)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "325a7ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get sequences from wipo database\n",
    "def get_seq_wipo(pat_no):\n",
    "    url=\"https://patentscope.wipo.int/search/en/search.jsf\"\n",
    "    driver.get(url)\n",
    "    \n",
    "    time.sleep(5)\n",
    "    \n",
    "    search_box=driver.find_element_by_id('simpleSearchForm:fpSearch:input')\n",
    "    search_box.send_keys(pat_no)\n",
    "    \n",
    "    #WO2018081194\n",
    "    search_box.send_keys(Keys.ENTER)\n",
    "    \n",
    "    time.sleep(5)\n",
    "    \n",
    "    # find the column having documents\n",
    "    m=0\n",
    "    ts=driver.find_elements_by_xpath('//*[@id=\"detailMainForm:MyTabViewId\"]/ul/li')\n",
    "    for i in range(len(ts)):\n",
    "        if(ts[i].text=='Documents'):\n",
    "            m=i+1\n",
    "            \n",
    "    try:\n",
    "        #select the documents\n",
    "        documents=driver.find_element_by_xpath('//*[@id=\"detailMainForm:MyTabViewId\"]/ul/li['+str(m)+']/a')\n",
    "        documents.send_keys(Keys.ENTER)\n",
    "    except Exception:\n",
    "        pass\n",
    "    \n",
    "    time.sleep(5)\n",
    "    \n",
    "    x=driver.find_elements_by_class_name('ui-datatable')\n",
    "    iD=x[1].get_attribute('id') #get the second table\n",
    "    \n",
    "    p='//*[@id=\"'+iD+'\"]/div[2]/table/tbody/tr' #rows of second table\n",
    "    table=driver.find_elements_by_xpath(p)\n",
    "    \n",
    "    #find the row in which sequence listing text file is present\n",
    "    fnd=0\n",
    "    for i in range(len(table)):\n",
    "        s=table[i].text\n",
    "        if(s.find('Sequence Listing')!=-1 and s.find('TXT')!=-1):\n",
    "            fnd=i+1\n",
    "            break\n",
    "    \n",
    "    time.sleep(5)\n",
    "    \n",
    "    #download the text file\n",
    "    sind=driver.find_element_by_xpath('//*[@id=\"'+iD+'\"]/div[2]/table/tbody/'+'/tr['+str(fnd)+']/td[3]/div/span[1]/a')\n",
    "    sind.send_keys(Keys.ENTER)\n",
    "    \n",
    "    tesxt=driver.find_element_by_tag_name('body')\n",
    "    file=tesxt.text.split('\\n')\n",
    "    df=get_data(file)\n",
    "    \n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "34854aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dicts={'Ala':'a', 'Asx':'b', 'Cys':'c',\n",
    "       'Asp':'d', 'Glu':'e', 'Phe':'f',\n",
    "       'Gly':'g', 'His':'h', 'Ile':'i',\n",
    "       'Lys':'k', 'Leu':'l', 'Met':'m',\n",
    "       'Asn':'n', 'Pro':'p', 'Gln':'q',\n",
    "       'Arg':'r', 'Ser':'s', 'Thr':'t',\n",
    "       'Sec':'u', 'Val':'v', 'Trp':'w',\n",
    "       'Xaa':'x', 'Tyr':'y', 'Glx':'z'}\n",
    "\n",
    "lis_amin=[]\n",
    "for key in dicts:\n",
    "    lis_amin.append(key)\n",
    "    \n",
    "lis_chk=['a', 't', 'g', 'c']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ca7f8c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get sequences for us patents from website \"https://appft1.uspto.gov/netahtml/PTO/srchnum.html\"\n",
    "def get_us_patents_1(patent_no):\n",
    "    url=\"https://appft1.uspto.gov/netahtml/PTO/srchnum.html\"\n",
    "    driver.get(url)\n",
    "\n",
    "    search_box=driver.find_element_by_xpath('//*[@id=\"qry\"]')\n",
    "    search_box.send_keys(patent_no)\n",
    "    #20090023654\n",
    "    search_box.send_keys(Keys.ENTER)\n",
    "\n",
    "    lnk=driver.find_element_by_xpath('/html/body/table/tbody/tr[2]/td[3]/a')\n",
    "    driver.get(lnk.get_attribute('href'))\n",
    "\n",
    "    el = driver.find_elements_by_tag_name('p')\n",
    "    ans=-1\n",
    "    for i in range(len(el)):\n",
    "        tsxt=el[i].text\n",
    "        if(tsxt.find('Sequence CWU')!=-1):\n",
    "            ans=i\n",
    "            break\n",
    "\n",
    "    import re\n",
    "    data=el[ans].text    \n",
    "\n",
    "\n",
    "    name=[]\n",
    "    new_data=[]\n",
    "    res = re.split('DNA|PRT', data)\n",
    "    for i in range(1,len(res)):\n",
    "        for j in range(len(res[i])-2):\n",
    "            if(res[i][j].isdigit()==True and res[i][j+1].isalpha()==True): #pattern number followed by alphabet\n",
    "                \n",
    "                if(res[i][j+1].isupper()):\n",
    "                    cnt=0\n",
    "                    if res[i][j+1:j+4] in lis_amin: # Check if three letter word is amino acid from dicts\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "                    \n",
    "                else:\n",
    "                    cnt=0\n",
    "                    for k in range(j+1,j+10):\n",
    "                        if(res[i][k]!=' '):\n",
    "                            if res[i][k] not in lis_chk: #check if 10 letter word contains 'a','t','g','c'\n",
    "                                cnt+=1\n",
    "                                break\n",
    "                    if(cnt==0):\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "            elif(res[i][j].isdigit()==True and res[i][j+2].isalpha()==True): #if the pattern is changed number followed by space then alphabet\n",
    "                \n",
    "                \n",
    "                if(res[i][j+2].isupper()):\n",
    "                    cnt=0\n",
    "                    if res[i][j+2:j+5] in lis_amin:\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "                    \n",
    "                else:\n",
    "                    cnt=0\n",
    "                    for k in range(j+2,j+10):\n",
    "                        if(res[i][k]!=' '):\n",
    "                            if res[i][k] not in lis_chk:\n",
    "                                cnt+=1\n",
    "                                break\n",
    "                    if(cnt==0):\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "                                \n",
    "                \n",
    "\n",
    "    fin_data=[]\n",
    "    for i in range(len(new_data)):\n",
    "        fin_data.append(re.sub(r'[^a-zA-Z ]+', '', new_data[i])) #clean the data\n",
    "\n",
    "    seq=[]\n",
    "    for i in range(len(fin_data)):\n",
    "        ix=fin_data[i].find(\"SEQ ID NO\") #remove everything after 'SEQ ID NO' if present\n",
    "        if(ix!=-1):\n",
    "            fin_data[i]=fin_data[i][:ix]\n",
    "        x=fin_data[i].split()\n",
    "        temp=\"\"\n",
    "        if(fin_data[i].islower()):\n",
    "            for j in range(len(x)):\n",
    "                temp+=x[j]\n",
    "        else:\n",
    "            for j in range(len(x)):\n",
    "                try:    \n",
    "                    \n",
    "                    temp+=exp_to_short(x[j]).upper()\n",
    "                    \n",
    "                except:\n",
    "                      pass\n",
    "                  \n",
    "        seq.append(temp)\n",
    "        \n",
    "    dicts={\"Name\": name, \"Sequence\": seq}\n",
    "    df=pd.DataFrame(dicts)\n",
    "    return df\n",
    "    \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "69ecee47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get Us patents from the website \"https://patft.uspto.gov/netahtml/PTO/srchnum.htm\"\n",
    "def get_us_patents_2(patent_no):\n",
    "    url=\"https://patft.uspto.gov/netahtml/PTO/srchnum.htm\"\n",
    "    driver.get(url)\n",
    "\n",
    "    time.sleep(3)\n",
    "    search_box=driver.find_element_by_xpath('//*[@id=\"qry\"]')\n",
    "    search_box.send_keys(patent_no)\n",
    "    #20090023654\n",
    "    search_box.send_keys(Keys.ENTER)\n",
    "\n",
    "    time.sleep(3)\n",
    "    el = driver.find_element_by_tag_name('body')\n",
    "    el2=el.text\n",
    "    ans=el2.find('SEQUENCE LISTINGS')\n",
    "    '''for i in range(len(el2)):\n",
    "        if(el2[i]=='SEQUENCE LISTINGS'):\n",
    "            ans=i+1\n",
    "            break\n",
    "        \n",
    "    for i in range(ans,len(el2)):\n",
    "        if(len(el2[i])>10):\n",
    "            ans=i\n",
    "            break'''\n",
    "\n",
    "\n",
    "    data=el2[ans:] \n",
    "\n",
    "\n",
    "    name=[]\n",
    "    new_data=[]\n",
    "    res = re.split('DNA|PRT', data)\n",
    "    for i in range(1,len(res)):\n",
    "        for j in range(len(res[i])-2):\n",
    "            if(res[i][j].isdigit()==True and res[i][j+1].isalpha()==True):\n",
    "                \n",
    "                if(res[i][j+1].isupper()):\n",
    "                    cnt=0\n",
    "                    if res[i][j+1:j+4] in lis_amin:\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "                    \n",
    "                else:\n",
    "                    cnt=0\n",
    "                    for k in range(j+1,j+10):\n",
    "                        if(res[i][k]!=' '):\n",
    "                            if res[i][k] not in lis_chk:\n",
    "                                cnt+=1\n",
    "                                break\n",
    "                    if(cnt==0):\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "            elif(res[i][j].isdigit()==True and res[i][j+2].isalpha()==True):\n",
    "                \n",
    "                \n",
    "                if(res[i][j+2].isupper()):\n",
    "                    cnt=0\n",
    "                    if res[i][j+2:j+5] in lis_amin:\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "                    \n",
    "                else:\n",
    "                    cnt=0\n",
    "                    for k in range(j+2,j+10):\n",
    "                        if(res[i][k]!=' '):\n",
    "                            if res[i][k] not in lis_chk:\n",
    "                                cnt+=1\n",
    "                                break\n",
    "                    if(cnt==0):\n",
    "                        name.append(res[i][:j])\n",
    "                        new_data.append(res[i][j:])\n",
    "                        break\n",
    "                                \n",
    "                \n",
    "\n",
    "    fin_data=[]\n",
    "    for i in range(len(new_data)):\n",
    "        fin_data.append(re.sub(r'[^a-zA-Z ]+', '', new_data[i]))\n",
    "\n",
    "    seq=[]\n",
    "    for i in range(len(fin_data)):\n",
    "        #ix=fin_data[i].find(\"SEQ ID NO\")\n",
    "        #fin_data[i]=fin_data[i][:ix]\n",
    "        x=fin_data[i].split()\n",
    "        temp=\"\"\n",
    "        if(fin_data[i].islower()):\n",
    "            for j in range(len(x)):\n",
    "                temp+=x[j]\n",
    "        else:\n",
    "            for j in range(len(x)):\n",
    "                try:    \n",
    "                    \n",
    "                    temp+=exp_to_short(x[j]).upper()\n",
    "                    \n",
    "                except:\n",
    "                      pass\n",
    "                  \n",
    "        seq.append(temp)\n",
    "        \n",
    "    dicts={\"Name\": name, \"Sequence\": seq}\n",
    "    df=pd.DataFrame(dicts)\n",
    "    return df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b206cf1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "t=get_seq_wipo('WO2018081194')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "58a6af6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence_ID_number</th>\n",
       "      <th>Sequence_length</th>\n",
       "      <th>Sequence_type</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Other_info</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1578</td>\n",
       "      <td>DNA</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>atggaacagatttacaatggaatgctgccgcttggtcttagccaac...</td>\n",
       "      <td>MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1578</td>\n",
       "      <td>DNA</td>\n",
       "      <td>Artificial Sequence</td>\n",
       "      <td>derived from Holdemania massiliensis sequence</td>\n",
       "      <td>atggaacagatttataatggcatgttaccgctgggtttgtcccagc...</td>\n",
       "      <td>MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1578</td>\n",
       "      <td>DNA</td>\n",
       "      <td>Artificial Sequence</td>\n",
       "      <td>derived from Holdemania massiliensis</td>\n",
       "      <td>atggagcagatctacaatgggatgctgccactcggcctgagccagc...</td>\n",
       "      <td>MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>525</td>\n",
       "      <td>PRT</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...</td>\n",
       "      <td>MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>PRT</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>YGTHLLLDIRLGG</td>\n",
       "      <td>YGTHLLLDIRLGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>1611</td>\n",
       "      <td>DNA</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>atgaggctaaggagtacgaaaggagaaatcaaaatggaacagattt...</td>\n",
       "      <td>MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>1611</td>\n",
       "      <td>DNA</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>atgcgtcttcgttctactaaaggcgaaatcaagatggaacagattt...</td>\n",
       "      <td>MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>1611</td>\n",
       "      <td>DNA</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>atgcgccttcggagcacaaaaggagaaatcaagatggagcagatct...</td>\n",
       "      <td>MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>536</td>\n",
       "      <td>PRT</td>\n",
       "      <td>Holdemania massiliensis</td>\n",
       "      <td></td>\n",
       "      <td>MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...</td>\n",
       "      <td>MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sequence_ID_number Sequence_length Sequence_type                  Organism  \\\n",
       "0                  1            1578           DNA   Holdemania massiliensis   \n",
       "1                  2            1578           DNA       Artificial Sequence   \n",
       "2                  3            1578           DNA       Artificial Sequence   \n",
       "3                  4             525           PRT   Holdemania massiliensis   \n",
       "4                  5              13           PRT   Holdemania massiliensis   \n",
       "5                  6            1611           DNA   Holdemania massiliensis   \n",
       "6                  7            1611           DNA   Holdemania massiliensis   \n",
       "7                  8            1611           DNA   Holdemania massiliensis   \n",
       "8                  9             536           PRT   Holdemania massiliensis   \n",
       "\n",
       "                                       Other_info  \\\n",
       "0                                                   \n",
       "1   derived from Holdemania massiliensis sequence   \n",
       "2            derived from Holdemania massiliensis   \n",
       "3                                                   \n",
       "4                                                   \n",
       "5                                                   \n",
       "6                                                   \n",
       "7                                                   \n",
       "8                                                   \n",
       "\n",
       "                                            Sequence  \\\n",
       "0  atggaacagatttacaatggaatgctgccgcttggtcttagccaac...   \n",
       "1  atggaacagatttataatggcatgttaccgctgggtttgtcccagc...   \n",
       "2  atggagcagatctacaatgggatgctgccactcggcctgagccagc...   \n",
       "3  MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...   \n",
       "4                                      YGTHLLLDIRLGG   \n",
       "5  atgaggctaaggagtacgaaaggagaaatcaaaatggaacagattt...   \n",
       "6  atgcgtcttcgttctactaaaggcgaaatcaagatggaacagattt...   \n",
       "7  atgcgccttcggagcacaaaaggagaaatcaagatggagcagatct...   \n",
       "8  MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...   \n",
       "\n",
       "                                             Protein  \n",
       "0  MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...  \n",
       "1  MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...  \n",
       "2  MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...  \n",
       "3  MEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLLAEQSPTIIYDK...  \n",
       "4                                      YGTHLLLDIRLGG  \n",
       "5  MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...  \n",
       "6  MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...  \n",
       "7  MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...  \n",
       "8  MRLRSTKGEIKMEQIYNGMLPLGLSQRLNEPVMMLEKGEQAKPQLL...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "31180c46",
   "metadata": {},
   "outputs": [],
   "source": [
    "u=get_us_patents_1('20110154524')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "05a6af62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bacillus thuringiensisPEPTIDE(0)...(0)Cry1F</td>\n",
       "      <td>MENNIQNQCVPYNCLNNPEVEILNEERSTGRLPLDISLSLTRFLLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacillus thuringiensisPEPTIDE(0)...(0)Cry34Ab1</td>\n",
       "      <td>MSAREVHIDVNNKTGHTLQLEDKTKLDGGRWRTSPTNVANDQIKTF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacillus thuringiensisPEPTIDE(0)...(0)Cry35Ab1</td>\n",
       "      <td>MLDTNKVYEISNHANGLYAATYLSLDDSGVSLMNKNDDDIDDYNLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Streptomyces viridochromogenesPEPTIDE(0)...(0)...</td>\n",
       "      <td>MSPERRPVEIRPATAADMAAVCDIVNHYIETSTVNFRTEPQTPQEW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>region of Plasmid PHP27118</td>\n",
       "      <td>gtttacccgccaatatatcctgtcaaacactgatagtttaaacgct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Artificial SequenceThe complete sequence of th...</td>\n",
       "      <td>cgatctctgcagacgtttgaacaatggaagaaacacacctccacgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>agggcctcatatccggtctcagttac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>cacgctgcactgcaggcatgcaagctggc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ccgctgtatcacaagggctggtacc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ggagcccgtgtagagcatgacgatc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>cgatctctgcagacgtttgaac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>gctgcactgcaggcatgc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ggtaattaatgtcatcctaaaacccctg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ggtcgacggatccttactcg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>tcctacgccactatcaacacc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ctcatgactcaggacttgtggc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>atcagcctctacttcgac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ctccatgatcttcgtctcatgtg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>caccaactccatccagaagtggc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>acaattcctacacctaagtgcacaac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>acctgcagaagcttcggtccg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>cgtgggaagtcatctgtgtacg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>gcagagcctttataagttgagcc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>ggataacatagtggatttgcattgtcc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>gacttaatttcttttagcataatgttattaaag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Artificial Sequencemisc_feature(0)...(0)Primer...</td>\n",
       "      <td>caaactctggggctggatc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Artificial Sequence5' Junction Sequence of eve...</td>\n",
       "      <td>ctcttcaggatgatagttta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Artificial Sequence3' Junction Sequence of eve...</td>\n",
       "      <td>aatatatcctaagttgcttc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Name  \\\n",
       "0        Bacillus thuringiensisPEPTIDE(0)...(0)Cry1F    \n",
       "1     Bacillus thuringiensisPEPTIDE(0)...(0)Cry34Ab1    \n",
       "2     Bacillus thuringiensisPEPTIDE(0)...(0)Cry35Ab1    \n",
       "3   Streptomyces viridochromogenesPEPTIDE(0)...(0)...   \n",
       "4                         region of Plasmid PHP27118    \n",
       "5   Artificial SequenceThe complete sequence of th...   \n",
       "6   Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "7   Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "8   Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "9   Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "10  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "11  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "12  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "13  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "14  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "15  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "16  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "17  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "18  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "19  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "20  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "21  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "22  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "23  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "24  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "25  Artificial Sequencemisc_feature(0)...(0)Primer...   \n",
       "26  Artificial Sequence5' Junction Sequence of eve...   \n",
       "27  Artificial Sequence3' Junction Sequence of eve...   \n",
       "\n",
       "                                             Sequence  \n",
       "0   MENNIQNQCVPYNCLNNPEVEILNEERSTGRLPLDISLSLTRFLLS...  \n",
       "1   MSAREVHIDVNNKTGHTLQLEDKTKLDGGRWRTSPTNVANDQIKTF...  \n",
       "2   MLDTNKVYEISNHANGLYAATYLSLDDSGVSLMNKNDDDIDDYNLK...  \n",
       "3   MSPERRPVEIRPATAADMAAVCDIVNHYIETSTVNFRTEPQTPQEW...  \n",
       "4   gtttacccgccaatatatcctgtcaaacactgatagtttaaacgct...  \n",
       "5   cgatctctgcagacgtttgaacaatggaagaaacacacctccacgt...  \n",
       "6                          agggcctcatatccggtctcagttac  \n",
       "7                       cacgctgcactgcaggcatgcaagctggc  \n",
       "8                           ccgctgtatcacaagggctggtacc  \n",
       "9                           ggagcccgtgtagagcatgacgatc  \n",
       "10                             cgatctctgcagacgtttgaac  \n",
       "11                                 gctgcactgcaggcatgc  \n",
       "12                       ggtaattaatgtcatcctaaaacccctg  \n",
       "13                               ggtcgacggatccttactcg  \n",
       "14                              tcctacgccactatcaacacc  \n",
       "15                             ctcatgactcaggacttgtggc  \n",
       "16                                 atcagcctctacttcgac  \n",
       "17                            ctccatgatcttcgtctcatgtg  \n",
       "18                            caccaactccatccagaagtggc  \n",
       "19                         acaattcctacacctaagtgcacaac  \n",
       "20                              acctgcagaagcttcggtccg  \n",
       "21                             cgtgggaagtcatctgtgtacg  \n",
       "22                            gcagagcctttataagttgagcc  \n",
       "23                        ggataacatagtggatttgcattgtcc  \n",
       "24                  gacttaatttcttttagcataatgttattaaag  \n",
       "25                                caaactctggggctggatc  \n",
       "26                               ctcttcaggatgatagttta  \n",
       "27                               aatatatcctaagttgcttc  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "03972e1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "u2=get_us_patents_2('9382550')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ce091519",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Artificial SequenceNucleotide sequence compris...</td>\n",
       "      <td>cttttatcatataagtgcctatataaataaagtttagtcagtcaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Artificial SequenceNucleotide sequence compris...</td>\n",
       "      <td>atagcgcgcaaactaggataaattatcgcgcgcggtgtcatctatc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI057</td>\n",
       "      <td>atagcgcgcaaactagga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer MAE115</td>\n",
       "      <td>tcggcaatgttatcggcatc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer HVH022</td>\n",
       "      <td>ccactgcttggcatttgag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI065</td>\n",
       "      <td>gggccggataaaattagcct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer HVH024</td>\n",
       "      <td>gtcattgtagggagtttgtc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer DPA312</td>\n",
       "      <td>tgcctcttgaactgtagc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI066</td>\n",
       "      <td>agataaaatcgtcagtgctg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI047</td>\n",
       "      <td>tgcctcttgaactgtagc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI048</td>\n",
       "      <td>acttgcagttgctgatgatg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>comprising EE-GH5 using GHI047 and GHI048</td>\n",
       "      <td>tgcctcttgaactgtagccacaacaccactttgaaacctcaaatct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI001</td>\n",
       "      <td>aacctaggctgctgaaggagc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer GHI002</td>\n",
       "      <td>caactcctccagtcatctccg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Artificial SequenceNucleotide sequence of the ...</td>\n",
       "      <td>agtttatatttcttacttttatcatataagtgcctatataaataaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Artificial SequenceNucleotide sequence compris...</td>\n",
       "      <td>aacaaataagccaattccaaagagaggatagctgatcttttcacta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer HVH036</td>\n",
       "      <td>caggtgttaggagcacaatc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer SHA028</td>\n",
       "      <td>cacaacaccactttgaaacctcaa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Artificial SequenceOligonucleotide primer MAE067</td>\n",
       "      <td>ccattcgattggaggttga</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Name  \\\n",
       "0   Artificial SequenceNucleotide sequence compris...   \n",
       "1   Artificial SequenceNucleotide sequence compris...   \n",
       "2   Artificial SequenceOligonucleotide primer GHI057    \n",
       "3   Artificial SequenceOligonucleotide primer MAE115    \n",
       "4   Artificial SequenceOligonucleotide primer HVH022    \n",
       "5   Artificial SequenceOligonucleotide primer GHI065    \n",
       "6   Artificial SequenceOligonucleotide primer HVH024    \n",
       "7   Artificial SequenceOligonucleotide primer DPA312    \n",
       "8   Artificial SequenceOligonucleotide primer GHI066    \n",
       "9   Artificial SequenceOligonucleotide primer GHI047    \n",
       "10  Artificial SequenceOligonucleotide primer GHI048    \n",
       "11         comprising EE-GH5 using GHI047 and GHI048    \n",
       "12  Artificial SequenceOligonucleotide primer GHI001    \n",
       "13  Artificial SequenceOligonucleotide primer GHI002    \n",
       "14  Artificial SequenceNucleotide sequence of the ...   \n",
       "15  Artificial SequenceNucleotide sequence compris...   \n",
       "16  Artificial SequenceOligonucleotide primer HVH036    \n",
       "17  Artificial SequenceOligonucleotide primer SHA028    \n",
       "18  Artificial SequenceOligonucleotide primer MAE067    \n",
       "\n",
       "                                             Sequence  \n",
       "0   cttttatcatataagtgcctatataaataaagtttagtcagtcaaa...  \n",
       "1   atagcgcgcaaactaggataaattatcgcgcgcggtgtcatctatc...  \n",
       "2                                  atagcgcgcaaactagga  \n",
       "3                                tcggcaatgttatcggcatc  \n",
       "4                                 ccactgcttggcatttgag  \n",
       "5                                gggccggataaaattagcct  \n",
       "6                                gtcattgtagggagtttgtc  \n",
       "7                                  tgcctcttgaactgtagc  \n",
       "8                                agataaaatcgtcagtgctg  \n",
       "9                                  tgcctcttgaactgtagc  \n",
       "10                               acttgcagttgctgatgatg  \n",
       "11  tgcctcttgaactgtagccacaacaccactttgaaacctcaaatct...  \n",
       "12                              aacctaggctgctgaaggagc  \n",
       "13                              caactcctccagtcatctccg  \n",
       "14  agtttatatttcttacttttatcatataagtgcctatataaataaa...  \n",
       "15  aacaaataagccaattccaaagagaggatagctgatcttttcacta...  \n",
       "16                               caggtgttaggagcacaatc  \n",
       "17                           cacaacaccactttgaaacctcaa  \n",
       "18                                ccattcgattggaggttga  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d857ed27",
   "metadata": {},
   "source": [
    "Get all the sequences for all the patents from the sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f23ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_file=pd.read_excel(\"patent_data.xlsx\")\n",
    "df1=pd.DataFrame()\n",
    "df2=pd.DataFrame()\n",
    "dint_wrk=[]\n",
    "wrk=[]\n",
    "\n",
    "for i in range(data_file.shape[0]):\n",
    "    \n",
    "    \n",
    "    if(data_file['Publication_No'][i][0]=='W'):\n",
    "        \n",
    "        try:\n",
    "            '''if(loaded_list[i][2:].lower().islower()):\n",
    "                dwnload_sequence(loaded_list[i][:-2])\n",
    "            else:\n",
    "                dwnload_sequence(loaded_list[i])'''\n",
    "            \n",
    "            if(data_file['Publication_No'][i][2:].lower().islower()):\n",
    "                df_data=get_seq_wipo(data_file['Publication_No'][i][:-2])\n",
    "                pat_nam=[data_file['Publication_No'][i]]*df_data.shape[0]\n",
    "                df_data['patent_number']=pat_nam\n",
    "                df1=pd.concat((df1,df_data))\n",
    "            \n",
    "            else:\n",
    "                df_data=get_seq_wipo(data_file['Publication_No'][i])\n",
    "                pat_nam=[data_file['Publication_No'][i]]*df_data.shape[0]\n",
    "                df_data['patent_number']=pat_nam\n",
    "                df1=pd.concat((df1,df_data))\n",
    "            \n",
    "            \n",
    "            wrk.append(data_file['Publication_No'][i])\n",
    "            if(i%200==0):\n",
    "                df1.to_csv(\"fetched_wipo\"+str(i)+\".csv\")\n",
    "                df2.to_csv(\"fetched_uspto\"+str(i)+\".csv\")\n",
    "                dn=0\n",
    "                \n",
    "                open_file = open(\"worked\"+str(i), \"wb\")\n",
    "                pickle.dump(wrk, open_file)\n",
    "                open_file.close()\n",
    "                \n",
    "                \n",
    "                open_file = open(\"dintworked\"+str(i), \"wb\")\n",
    "                pickle.dump(dint_wrk, open_file)\n",
    "                open_file.close()\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        except:\n",
    "            dint_wrk.append(data_file['Publication_No'][i])\n",
    "    \n",
    "    else:\n",
    "        try:\n",
    "            if(data_file['Publication_No'][i][2:].lower().islower()):\n",
    "                df_data=get_us_patents_1(data_file['Publication_No'][i][2:-2])\n",
    "                pat_nam=[data_file['Publication_No'][i]]*df_data.shape[0]\n",
    "                df_data['patent_number']=pat_nam\n",
    "                df2=pd.concat((df2,df_data))\n",
    "            else:\n",
    "                df_data=get_us_patents_1(data_file['Publication_No'][i][2:])\n",
    "                pat_nam=[data_file['Publication_No'][i]]*df_data.shape[0]\n",
    "                df_data['patent_number']=pat_nam\n",
    "                df2=pd.concat((df2,df_data))\n",
    "            \n",
    "            \n",
    "            wrk.append(data_file['Publication_No'][i])\n",
    "            if(i%200==0):\n",
    "                df1.to_csv(\"fetched_wipo\"+str(i)+\".csv\")\n",
    "                df2.to_csv(\"fetched_uspto\"+str(i)+\".csv\")\n",
    "                dn=0\n",
    "                \n",
    "                open_file = open(\"worked\"+str(i), \"wb\")\n",
    "                pickle.dump(wrk, open_file)\n",
    "                open_file.close()\n",
    "                \n",
    "                \n",
    "                open_file = open(\"dintworked\"+str(i), \"wb\")\n",
    "                pickle.dump(dint_wrk, open_file)\n",
    "                open_file.close()\n",
    "        \n",
    "        except:\n",
    "            try:\n",
    "                if(data_file['Publication_No'][i][2:].lower().islower()):\n",
    "                    df_data=get_us_patents_2(data_file['Publication_No'][i][2:-2])\n",
    "                    pat_nam=[data_file['Publication_No'][i]]*df_data.shape[0]\n",
    "                    df_data['patent_number']=pat_nam\n",
    "                    df2=pd.concat((df2,df_data))\n",
    "                else:\n",
    "                    df_data=get_us_patents_2(data_file['Publication_No'][i][2:])\n",
    "                    pat_nam=[data_file['Publication_No'][i]]*df_data.shape[0]\n",
    "                    df_data['patent_number']=pat_nam\n",
    "                    df2=pd.concat((df2,df_data))\n",
    "                    \n",
    "                wrk.append(data_file['Publication_No'][i])\n",
    "                if(i%200==0):\n",
    "                    df1.to_csv(\"fetched_wipo\"+str(i)+\".csv\")\n",
    "                    df2.to_csv(\"fetched_uspto\"+str(i)+\".csv\")\n",
    "                    dn=0\n",
    "                    \n",
    "                    open_file = open(\"worked\"+str(i), \"wb\")\n",
    "                    pickle.dump(wrk, open_file)\n",
    "                    open_file.close()\n",
    "                    \n",
    "                    \n",
    "                    open_file = open(\"dintworked\"+str(i), \"wb\")\n",
    "                    pickle.dump(dint_wrk, open_file)\n",
    "                    open_file.close()\n",
    "            \n",
    "            except:\n",
    "                dint_wrk.append(data_file['Publication_No'][i])\n",
    "                \n",
    "open_file = open(\"worked\"+str(i), \"wb\")\n",
    "pickle.dump(wrk, open_file)\n",
    "open_file.close()\n",
    "\n",
    "\n",
    "open_file = open(\"dintworked\"+str(i), \"wb\")\n",
    "pickle.dump(dint_wrk, open_file)\n",
    "open_file.close()\n",
    "                \n",
    "df1.to_csv('fetched_wipo_all.csv')\n",
    "df2.to_csv('fetched_uspto_all.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
